### PR TITLE
Support deferred procedures

### PR DESCRIPTION
**Pull request by MatthewHambley**
Adds support for the "deferred" keyword on type bound procedures.

### CODE REVIEW

File: src/fparser/Fortran2003.py [hunk 1]
```diff
@@ -2089,7 +2089,7 @@ def match(string):
         if line.startswith('('):
             i = line.find(')')
             if i==-1: return
-            iname = Interface_Name(line[1:-1].strip())
+            iname = Interface_Name(line[1:i].strip())
```
Comment (arporter) [inline src/fparser/Fortran2003.py]: Bug fix!

File: src/fparser/Fortran2003.py [hunk 2]
```diff
@@ -2154,10 +2154,11 @@ class Binding_Attr(STRINGBase): # R453
     <binding-attr> = PASS [ ( <arg-name> ) ]
                      | NOPASS
                      | NON_OVERRIDABLE
+                     | DEFERRED
                      | <access-spec>
     """
     subclass_names = ['Access_Spec', 'Binding_PASS_Arg_Name']
-    def match(string): return STRINGBase.match(['PASS', 'NOPASS', 'NON_OVERRIDABLE'], string)
+    def match(string): return STRINGBase.match(['PASS', 'NOPASS', 'NON_OVERRIDABLE', 'DEFERRED'], string)
```
Comment (arporter) [inline src/fparser/Fortran2003.py]: Include 'DEFERRED' as valid binding attribute. Please tidy this line for pep8/pylint (the `return` should be on a new line).

File: src/fparser/tests/fparser2/test_deferred_method.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+##############################################################################
+'''
```
Comment (arporter) [inline src/fparser/tests/fparser2/test_deferred_method.py]: Please add yourself as the author of this file.

Comment (MatthewHambley) [reply src/fparser/tests/fparser2/test_deferred_method.py]: Do you have a template for this as I can find no example of it in a sampling of other source files.

**Comment by arporter** [comment]
Hi @MatthewHambley, nothing fancy. Just something like:

    # Modified M. Hambley, UK Met Office

i.e. as Iva has done in the PSyclone source.

**Pull request merged**